File: android/src/main/java/com/getcapacitor/community/admob/nativeads/NativeAdExecutor.java
package com.getcapacitor.community.admob.nativeads;

import android.app.Activity;
import android.content.Context;
import android.view.View;
import android.view.ViewGroup;
import android.widget.FrameLayout;
import androidx.annotation.NonNull;
import androidx.core.util.Supplier;
import com.getcapacitor.JSArray;
import com.getcapacitor.JSObject;
import com.getcapacitor.PluginCall;
import com.getcapacitor.community.admob.helpers.RequestHelper;
import com.getcapacitor.community.admob.models.AdMobRevenueData;
import com.getcapacitor.community.admob.models.AdOptions;
import com.getcapacitor.community.admob.models.Executor;
import com.google.android.gms.ads.AdListener;
import com.google.android.gms.ads.AdLoader;
import com.google.android.gms.ads.LoadAdError;
import com.google.android.gms.ads.nativead.NativeAd;
import com.google.android.gms.ads.nativead.NativeAdOptions;
import com.google.android.gms.ads.nativead.NativeAdView;
import com.google.android.gms.common.util.BiConsumer;
import java.util.HashMap;
import java.util.Map;
import java.util.concurrent.atomic.AtomicBoolean;
import org.json.JSONException;
import org.json.JSONObject;

public class NativeAdExecutor extends Executor {

    public static final String NATIVE_TESTER_ID = "ca-app-pub-3940256099942544/2247696110";
    private static final int MAX_ADS_PER_FEED = 3;

    private static final class NativeAdState {

        final String feedId;
        final boolean isSmall;
        final Object loadToken;
        final NativeAd nativeAd;
        final NativeAdView adView;
        final FrameLayout clippingContainer;
        int generation = -1;

        NativeAdState(
            String feedId,
            boolean isSmall,
            Object loadToken,
            NativeAd nativeAd,
            NativeAdView adView,
            FrameLayout clippingContainer
        ) {
            this.feedId = feedId;
            this.isSmall = isSmall;
            this.loadToken = loadToken;
            this.nativeAd = nativeAd;
            this.adView = adView;
            this.clippingContainer = clippingContainer;
        }
    }

    private static final class PendingLoad {

        final PluginCall call;
        final String feedId;
        final Object token = new Object();

        PendingLoad(PluginCall call, String feedId) {
            this.call = call;
            this.feedId = feedId;
        }
    }

    private final Supplier<View> webViewSupplier;
    private final Map<String, NativeAdState> states = new HashMap<>();
    private final Map<String, PendingLoad> pendingLoads = new HashMap<>();
    private final NativeAdFeedSessions feedSessions = new NativeAdFeedSessions();

    public NativeAdExecutor(
        Supplier<Context> contextSupplier,
        Supplier<Activity> activitySupplier,
        Supplier<View> webViewSupplier,
        BiConsumer<String, JSObject> notifyListenersFunction,
        String pluginLogTag
    ) {
        super(contextSupplier, activitySupplier, notifyListenersFunction, pluginLogTag, "NativeAdExecutor");
        this.webViewSupplier = webViewSupplier;
    }

    public void startFeed(final PluginCall call) {
        final String feedId = requiredString(call, "feedId");
        final String sessionId = requiredString(call, "sessionId");
        if (feedId == null || sessionId == null) return;
        activitySupplier
            .get()
            .runOnUiThread(() -> {
                if (!feedSessions.isCurrent(feedId, sessionId)) clearFeed(feedId);
                if (!feedSessions.start(feedId, sessionId)) {
                    call.reject("At most two native ad feeds can be active");
                    return;
                }
                call.resolve();
            });
    }

    public void destroyFeed(final PluginCall call) {
        final String feedId = requiredString(call, "feedId");
        final String sessionId = requiredString(call, "sessionId");
        if (feedId == null || sessionId == null) return;
        activitySupplier
            .get()
            .runOnUiThread(() -> {
                if (feedSessions.isCurrent(feedId, sessionId)) {
                    clearFeed(feedId);
                    feedSessions.remove(feedId);
                }
                call.resolve();
            });
    }

    public void load(final PluginCall call) {
        final String feedId = requiredString(call, "feedId");
        final String sessionId = requiredString(call, "sessionId");
        final String slotKey = requiredString(call, "slotKey");
        if (feedId == null || sessionId == null || slotKey == null) return;

        final AdOptions options = AdOptions.getFactory().createGenericOptions(call, NATIVE_TESTER_ID);
        final String template = call.getString("template", "medium");
        final JSObject style = call.getObject("style", new JSObject());
        final String stateKey = stateKey(feedId, slotKey);
        final AtomicBoolean finished = new AtomicBoolean(false);
        final PendingLoad pendingLoad = new PendingLoad(call, feedId);

        activitySupplier
            .get()
            .runOnUiThread(() -> {
                if (!isCurrentSession(feedId, sessionId)) {
                    call.reject("Native ad feed session is no longer active");
                    return;
                }
                destroyState(stateKey);
                cancelPendingLoad(stateKey);
                long feedAdCount = states
                    .values()
                    .stream()
                    .filter((state) -> state.feedId.equals(feedId))
                    .count();
                feedAdCount += pendingLoads
                    .values()
                    .stream()
                    .filter((pending) -> pending.feedId.equals(feedId))
                    .count();
                if (feedAdCount >= MAX_ADS_PER_FEED) {
                    call.reject("At most three native ads can be active in a feed");
                    return;
                }
                pendingLoads.put(stateKey, pendingLoad);
                final AdLoader adLoader = new AdLoader.Builder(contextSupplier.get(), NATIVE_TESTER_ID)
                    .forNativeAd((nativeAd) -> {
                        if (pendingLoads.get(stateKey) != pendingLoad || !finished.compareAndSet(false, true)) {
                            nativeAd.destroy();
                            return;
                        }
                        pendingLoads.remove(stateKey);
                        NativeAdView adView = PluginNativeAdView.create(contextSupplier.get(), nativeAd, template, style);
                        FrameLayout clippingContainer = new FrameLayout(contextSupplier.get());
                        clippingContainer.setClipChildren(true);
                        clippingContainer.setClipToPadding(true);
                        clippingContainer.setVisibility(View.GONE);
                        clippingContainer.addView(adView);

                        ViewGroup overlayParent = getOverlayParent();
                        if (overlayParent == null) {
                            adView.destroy();
                            nativeAd.destroy();
                            call.reject("AdMob WebView parent is unavailable");
                            return;
                        }

                        NativeAdState state = new NativeAdState(
                            feedId,
                            "small".equals(template),
                            pendingLoad.token,
                            nativeAd,
                            adView,
                            clippingContainer
                        );
                        states.put(stateKey, state);
                        overlayParent.addView(clippingContainer);

                        nativeAd.setOnPaidEventListener((adValue) -> {
                            if (states.get(stateKey) != state) return;
                            String networkName = "";
                            String impressionId = "";
                            if (nativeAd.getResponseInfo() != null) {
                                networkName = nativeAd.getResponseInfo().getMediationAdapterClassName();
                                if (networkName == null) networkName = "";
                                impressionId = nativeAd.getResponseInfo().getResponseId();
                                if (impressionId == null) impressionId = "";
                            }
                            AdMobRevenueData data = new AdMobRevenueData(adValue, NATIVE_TESTER_ID, networkName, impressionId);
                            data.put("feedId", feedId);
                            data.put("slotKey", slotKey);
                            data.put("sessionId", sessionId);
                            notifyListeners(NativeAdPluginEvents.AD_PAID.getWebEventName(), data);
                        });

                        notifyListeners(NativeAdPluginEvents.LOADED.getWebEventName(), identity(feedId, sessionId, slotKey));
                        call.resolve();
                    })
                    .withNativeAdOptions(new NativeAdOptions.Builder().setAdChoicesPlacement(NativeAdOptions.ADCHOICES_TOP_RIGHT).build())
                    .withAdListener(
                        new AdListener() {
                            @Override
                            public void onAdFailedToLoad(@NonNull LoadAdError error) {
                                if (pendingLoads.get(stateKey) != pendingLoad) return;
                                pendingLoads.remove(stateKey);
                                JSObject event = identity(feedId, sessionId, slotKey);
                                event.put("code", error.getCode());
                                event.put("message", error.getMessage());
                                notifyListeners(NativeAdPluginEvents.FAILED_TO_LOAD.getWebEventName(), event);
                                if (finished.compareAndSet(false, true)) {
                                    call.reject(error.getMessage(), String.valueOf(error.getCode()), event);
                                }
                            }

                            @Override
                            public void onAdClicked() {
                                if (!isCurrent(stateKey, pendingLoad.token)) return;
                                notifyListeners(NativeAdPluginEvents.CLICKED.getWebEventName(), identity(feedId, sessionId, slotKey));
                            }

                            @Override
                            public void onAdImpression() {
                                if (!isCurrent(stateKey, pendingLoad.token)) return;
                                notifyListeners(NativeAdPluginEvents.IMPRESSION.getWebEventName(), identity(feedId, sessionId, slotKey));
                            }

                            @Override
                            public void onAdOpened() {
                                if (!isCurrent(stateKey, pendingLoad.token)) return;
                                notifyListeners(NativeAdPluginEvents.OPENED.getWebEventName(), identity(feedId, sessionId, slotKey));
                            }

                            @Override
                            public void onAdClosed() {
                                if (!isCurrent(stateKey, pendingLoad.token)) return;
                                notifyListeners(NativeAdPluginEvents.CLOSED.getWebEventName(), identity(feedId, sessionId, slotKey));
                            }
                        }
                    )
                    .build();

                adLoader.loadAd(RequestHelper.createRequest(options));
            });
    }

    public void updatePlacements(final PluginCall call) {
        final String feedId = requiredString(call, "feedId");
        final String sessionId = requiredString(call, "sessionId");
        if (feedId == null || sessionId == null) return;
        final long sequence = call.getLong("sequence", -1L);
        final JSArray placements = call.getArray("placements", new JSArray());
        activitySupplier
            .get()
            .runOnUiThread(() -> {
                if (!isCurrentSession(feedId, sessionId)) {
                    call.resolve();
                    return;
                }
                if (!feedSessions.accepts(feedId, sequence)) {
                    call.resolve();
                    return;
                }
                for (NativeAdState state : states.values()) {
                    if (state.feedId.equals(feedId)) state.clippingContainer.setVisibility(View.GONE);
                }
                for (int index = 0; index < placements.length(); index++) {
                    try {
                        applyPlacement(feedId, placements.getJSONObject(index));
                    } catch (JSONException ignored) {
                        // Invalid entries fail closed: every state was hidden above.
                    }
                }
                call.resolve();
            });
    }

    public void remove(final PluginCall call) {
        final String feedId = requiredString(call, "feedId");
        final String sessionId = requiredString(call, "sessionId");
        final String slotKey = requiredString(call, "slotKey");
        if (feedId == null || sessionId == null || slotKey == null) return;
        activitySupplier
            .get()
            .runOnUiThread(() -> {
                if (!isCurrentSession(feedId, sessionId)) {
                    call.resolve();
                    return;
                }
                cancelPendingLoad(stateKey(feedId, slotKey));
                destroyState(stateKey(feedId, slotKey));
                call.resolve();
            });
    }

    public void destroyAll() {
        for (String key : pendingLoads.keySet().toArray(new String[0])) cancelPendingLoad(key);
        for (String key : states.keySet().toArray(new String[0])) destroyState(key);
        feedSessions.clear();
    }

    private void applyPlacement(String expectedFeedId, JSONObject placement) throws JSONException {
        float density = contextSupplier.get().getResources().getDisplayMetrics().density;
        NativeAdPlacementValue value = NativeAdPlacementValue.parse(expectedFeedId, placement, density);
        if (value == null) return;
        NativeAdState state = states.get(stateKey(value.feedId, value.slotKey));
        if (state == null || value.generation < state.generation) return;
        int minimumWidth = Math.round((state.isSmall ? 120 : 144) * density);
        int minimumHeight = Math.round((state.isSmall ? 120 : 300) * density);
        if (value.rectWidth < minimumWidth || value.rectHeight < minimumHeight) return;

        View webView = webViewSupplier.get();
        if (webView == null) return;
        ViewGroup.LayoutParams layoutParams = state.clippingContainer.getLayoutParams();
        layoutParams.width = value.clipWidth;
        layoutParams.height = value.clipHeight;
        state.clippingContainer.setLayoutParams(layoutParams);
        state.clippingContainer.setX(webView.getX() + value.clipX);
        state.clippingContainer.setY(webView.getY() + value.clipY);

        state.adView.setLayoutParams(new FrameLayout.LayoutParams(value.rectWidth, value.rectHeight));
        state.adView.setX(value.rectX - value.clipX);
        state.adView.setY(value.rectY - value.clipY);
        state.generation = value.generation;
        state.clippingContainer.setVisibility(View.VISIBLE);
        state.clippingContainer.bringToFront();
    }

    private ViewGroup getOverlayParent() {
        View webView = webViewSupplier.get();
        return webView != null && webView.getParent() instanceof ViewGroup ? (ViewGroup) webView.getParent() : null;
    }

    private void destroyState(String stateKey) {
        NativeAdState state = states.remove(stateKey);
        if (state == null) return;
        ViewGroup parent = (ViewGroup) state.clippingContainer.getParent();
        if (parent != null) parent.removeView(state.clippingContainer);
        state.adView.destroy();
        state.nativeAd.destroy();
    }

    private void cancelPendingLoad(String stateKey) {
        PendingLoad pendingLoad = pendingLoads.remove(stateKey);
        if (pendingLoad != null) pendingLoad.call.reject("Native ad load was cancelled");
    }

    private boolean isCurrent(String stateKey, Object loadToken) {
        NativeAdState state = states.get(stateKey);
        return state != null && state.loadToken == loadToken;
    }

    private boolean isCurrentSession(String feedId, String sessionId) {
        return feedSessions.isCurrent(feedId, sessionId);
    }

    private void clearFeed(String feedId) {
        for (String key : pendingLoads.keySet().toArray(new String[0])) {
            PendingLoad pendingLoad = pendingLoads.get(key);
            if (pendingLoad != null && pendingLoad.feedId.equals(feedId)) cancelPendingLoad(key);
        }
        for (String key : states.keySet().toArray(new String[0])) {
            NativeAdState state = states.get(key);
            if (state != null && state.feedId.equals(feedId)) destroyState(key);
        }
    }

    private String requiredString(PluginCall call, String name) {
        String value = call.getString(name);
        if (value == null || value.trim().isEmpty()) {
            call.reject(name + " must not be empty");
            return null;
        }
        return value.trim();
    }

    private String stateKey(String feedId, String slotKey) {
        return feedId + "\u0000" + slotKey;
    }

    private JSObject identity(String feedId, String sessionId, String slotKey) {
        JSObject object = new JSObject();
        object.put("feedId", feedId);
        object.put("sessionId", sessionId);
        object.put("slotKey", slotKey);
        return object;
    }
}

File: android/src/main/java/com/getcapacitor/community/admob/nativeads/NativeAdExecutor.kt
package com.getcapacitor.community.admob.nativeads

import android.app.Activity
import android.content.Context
import android.view.View
import android.view.ViewGroup
import android.widget.FrameLayout
import androidx.core.util.Supplier
import com.getcapacitor.JSArray
import com.getcapacitor.JSObject
import com.getcapacitor.PluginCall
import com.getcapacitor.community.admob.helpers.RequestHelper
import com.getcapacitor.community.admob.models.AdMobRevenueData
import com.getcapacitor.community.admob.models.AdOptions
import com.getcapacitor.community.admob.models.Executor
import com.google.android.gms.ads.AdListener
import com.google.android.gms.ads.AdLoader
import com.google.android.gms.ads.LoadAdError
import com.google.android.gms.ads.nativead.NativeAd
import com.google.android.gms.ads.nativead.NativeAdOptions
import com.google.android.gms.ads.nativead.NativeAdView
import com.google.android.gms.common.util.BiConsumer
import org.json.JSONException
import org.json.JSONObject
import java.util.concurrent.atomic.AtomicBoolean
import kotlin.math.roundToInt

class NativeAdExecutor(
    contextSupplier: Supplier<Context>,
    activitySupplier: Supplier<Activity>,
    private val webViewSupplier: Supplier<View>,
    notifyListenersFunction: BiConsumer<String, JSObject>,
    pluginLogTag: String,
) : Executor(contextSupplier, activitySupplier, notifyListenersFunction, pluginLogTag, "NativeAdExecutor") {
    private data class NativeAdState(
        val feedId: String,
        val isSmall: Boolean,
        val loadToken: Any,
        val nativeAd: NativeAd,
        val adView: NativeAdView,
        val clippingContainer: FrameLayout,
        var generation: Int = -1,
    )

    private data class PendingLoad(
        val call: PluginCall,
        val feedId: String,
        val token: Any = Any(),
    )

    private val states = mutableMapOf<String, NativeAdState>()
    private val pendingLoads = mutableMapOf<String, PendingLoad>()
    private val feedSessions = NativeAdFeedSessions()

    fun startFeed(call: PluginCall) {
        val feedId = requiredString(call, "feedId") ?: return
        val sessionId = requiredString(call, "sessionId") ?: return
        activitySupplier.get().runOnUiThread {
            if (!feedSessions.isCurrent(feedId, sessionId)) clearFeed(feedId)
            if (!feedSessions.start(feedId, sessionId)) {
                call.reject("At most two native ad feeds can be active")
                return@runOnUiThread
            }
            call.resolve()
        }
    }

    fun destroyFeed(call: PluginCall) {
        val feedId = requiredString(call, "feedId") ?: return
        val sessionId = requiredString(call, "sessionId") ?: return
        activitySupplier.get().runOnUiThread {
            if (feedSessions.isCurrent(feedId, sessionId)) {
                clearFeed(feedId)
                feedSessions.remove(feedId)
            }
            call.resolve()
        }
    }

    fun load(call: PluginCall) {
        val feedId = requiredString(call, "feedId") ?: return
        val sessionId = requiredString(call, "sessionId") ?: return
        val slotKey = requiredString(call, "slotKey") ?: return
        val options = AdOptions.getFactory().createGenericOptions(call, NATIVE_TESTER_ID)
        val template = call.getString("template", "medium") ?: "medium"
        val style = call.getObject("style", JSObject()) ?: JSObject()
        val stateKey = stateKey(feedId, slotKey)
        val finished = AtomicBoolean(false)
        val pendingLoad = PendingLoad(call, feedId)

        activitySupplier.get().runOnUiThread {
            if (!isCurrentSession(feedId, sessionId)) {
                call.reject("Native ad feed session is no longer active")
                return@runOnUiThread
            }
            destroyState(stateKey)
            cancelPendingLoad(stateKey)
            val feedAdCount = states.values.count { it.feedId == feedId } + pendingLoads.values.count { it.feedId == feedId }
            if (feedAdCount >= MAX_ADS_PER_FEED) {
                call.reject("At most three native ads can be active in a feed")
                return@runOnUiThread
            }
            pendingLoads[stateKey] = pendingLoad
            val adLoader = AdLoader.Builder(contextSupplier.get(), NATIVE_TESTER_ID)
                .forNativeAd { nativeAd ->
                    handleLoadedAd(
                        nativeAd = nativeAd,
                        stateKey = stateKey,
                        pendingLoad = pendingLoad,
                        finished = finished,
                        feedId = feedId,
                        sessionId = sessionId,
                        slotKey = slotKey,
                        template = template,
                        style = style,
                    )
                }
                .withNativeAdOptions(
                    NativeAdOptions.Builder()
                        .setAdChoicesPlacement(NativeAdOptions.ADCHOICES_TOP_RIGHT)
                        .build(),
                )
                .withAdListener(adListener(stateKey, pendingLoad, finished, feedId, sessionId, slotKey))
                .build()

            adLoader.loadAd(RequestHelper.createRequest(options))
        }
    }

    fun updatePlacements(call: PluginCall) {
        val feedId = requiredString(call, "feedId") ?: return
        val sessionId = requiredString(call, "sessionId") ?: return
        val sequence = call.getLong("sequence", -1L) ?: -1L
        val placements = call.getArray("placements", JSArray()) ?: JSArray()
        activitySupplier.get().runOnUiThread {
            if (!isCurrentSession(feedId, sessionId) || !feedSessions.accepts(feedId, sequence)) {
                call.resolve()
                return@runOnUiThread
            }
            states.values.filter { it.feedId == feedId }.forEach { it.clippingContainer.visibility = View.GONE }
            repeat(placements.length()) { index ->
                try {
                    applyPlacement(feedId, placements.getJSONObject(index))
                } catch (_: JSONException) {
                    // Invalid entries fail closed: every state was hidden above.
                }
            }
            call.resolve()
        }
    }

    fun remove(call: PluginCall) {
        val feedId = requiredString(call, "feedId") ?: return
        val sessionId = requiredString(call, "sessionId") ?: return
        val slotKey = requiredString(call, "slotKey") ?: return
        activitySupplier.get().runOnUiThread {
            if (isCurrentSession(feedId, sessionId)) {
                val key = stateKey(feedId, slotKey)
                cancelPendingLoad(key)
                destroyState(key)
            }
            call.resolve()
        }
    }

    fun destroyAll() {
        pendingLoads.keys.toList().forEach(::cancelPendingLoad)
        states.keys.toList().forEach(::destroyState)
        feedSessions.clear()
    }

    private fun handleLoadedAd(
        nativeAd: NativeAd,
        stateKey: String,
        pendingLoad: PendingLoad,
        finished: AtomicBoolean,
        feedId: String,
        sessionId: String,
        slotKey: String,
        template: String,
        style: JSObject,
    ) {
        if (pendingLoads[stateKey] !== pendingLoad || !finished.compareAndSet(false, true)) {
            nativeAd.destroy()
            return
        }
        pendingLoads.remove(stateKey)
        val adView = PluginNativeAdView.create(contextSupplier.get(), nativeAd, template, style)
        val clippingContainer = FrameLayout(contextSupplier.get()).apply {
            clipChildren = true
            clipToPadding = true
            visibility = View.GONE
            addView(adView)
        }
        val overlayParent = overlayParent()
        if (overlayParent == null) {
            adView.destroy()
            nativeAd.destroy()
            pendingLoad.call.reject("AdMob WebView parent is unavailable")
            return
        }

        val state = NativeAdState(feedId, template == "small", pendingLoad.token, nativeAd, adView, clippingContainer)
        states[stateKey] = state
        overlayParent.addView(clippingContainer)
        nativeAd.setOnPaidEventListener { adValue ->
            if (states[stateKey] !== state) return@setOnPaidEventListener
            val responseInfo = nativeAd.responseInfo
            val data = AdMobRevenueData(
                adValue,
                NATIVE_TESTER_ID,
                responseInfo?.mediationAdapterClassName ?: "",
                responseInfo?.responseId ?: "",
            ).apply {
                put("feedId", feedId)
                put("slotKey", slotKey)
                put("sessionId", sessionId)
            }
            notifyListeners(NativeAdPluginEvents.AD_PAID, data)
        }

        notifyListeners(NativeAdPluginEvents.LOADED, identity(feedId, sessionId, slotKey))
        pendingLoad.call.resolve()
    }

    private fun adListener(
        stateKey: String,
        pendingLoad: PendingLoad,
        finished: AtomicBoolean,
        feedId: String,
        sessionId: String,
        slotKey: String,
    ): AdListener = object : AdListener() {
        override fun onAdFailedToLoad(error: LoadAdError) {
            if (pendingLoads[stateKey] !== pendingLoad) return
            pendingLoads.remove(stateKey)
            val event = identity(feedId, sessionId, slotKey).apply {
                put("code", error.code)
                put("message", error.message)
            }
            notifyListeners(NativeAdPluginEvents.FAILED_TO_LOAD, event)
            if (finished.compareAndSet(false, true)) {
                pendingLoad.call.reject(error.message, error.code.toString(), event)
            }
        }

        override fun onAdClicked() = notifyIfCurrent(NativeAdPluginEvents.CLICKED)

        override fun onAdImpression() = notifyIfCurrent(NativeAdPluginEvents.IMPRESSION)

        override fun onAdOpened() = notifyIfCurrent(NativeAdPluginEvents.OPENED)

        override fun onAdClosed() = notifyIfCurrent(NativeAdPluginEvents.CLOSED)

        private fun notifyIfCurrent(eventName: String) {
            if (isCurrent(stateKey, pendingLoad.token)) {
                notifyListeners(eventName, identity(feedId, sessionId, slotKey))
            }
        }
    }

    private fun applyPlacement(expectedFeedId: String, placement: JSONObject) {
        val density = contextSupplier.get().resources.displayMetrics.density
        val value = NativeAdPlacementValue.parse(expectedFeedId, placement, density) ?: return
        val state = states[stateKey(value.feedId, value.slotKey)] ?: return
        if (value.generation < state.generation) return
        val minimumWidth = (if (state.isSmall) 120 else 144) * density
        val minimumHeight = (if (state.isSmall) 120 else 300) * density
        if (value.rectWidth < minimumWidth.roundToInt() || value.rectHeight < minimumHeight.roundToInt()) return

        val webView = webViewSupplier.get() ?: return
        state.clippingContainer.layoutParams = state.clippingContainer.layoutParams.apply {
            width = value.clipWidth
            height = value.clipHeight
        }
        state.clippingContainer.x = webView.x + value.clipX
        state.clippingContainer.y = webView.y + value.clipY
        state.adView.layoutParams = FrameLayout.LayoutParams(value.rectWidth, value.rectHeight)
        state.adView.x = (value.rectX - value.clipX).toFloat()
        state.adView.y = (value.rectY - value.clipY).toFloat()
        state.generation = value.generation
        state.clippingContainer.visibility = View.VISIBLE
        state.clippingContainer.bringToFront()
    }

    private fun overlayParent(): ViewGroup? = webViewSupplier.get()?.parent as? ViewGroup

    private fun destroyState(stateKey: String) {
        val state = states.remove(stateKey) ?: return
        (state.clippingContainer.parent as? ViewGroup)?.removeView(state.clippingContainer)
        state.adView.destroy()
        state.nativeAd.destroy()
    }

    private fun cancelPendingLoad(stateKey: String) {
        pendingLoads.remove(stateKey)?.call?.reject("Native ad load was cancelled")
    }

    private fun isCurrent(stateKey: String, loadToken: Any): Boolean = states[stateKey]?.loadToken === loadToken

    private fun isCurrentSession(feedId: String, sessionId: String): Boolean = feedSessions.isCurrent(feedId, sessionId)

    private fun clearFeed(feedId: String) {
        pendingLoads.filterValues { it.feedId == feedId }.keys.toList().forEach(::cancelPendingLoad)
        states.filterValues { it.feedId == feedId }.keys.toList().forEach(::destroyState)
    }

    private fun requiredString(call: PluginCall, name: String): String? {
        val value = call.getString(name)?.trim()
        if (value.isNullOrEmpty()) {
            call.reject("$name must not be empty")
            return null
        }
        return value
    }

    private fun stateKey(feedId: String, slotKey: String): String = "$feedId\u0000$slotKey"

    private fun identity(feedId: String, sessionId: String, slotKey: String): JSObject = JSObject().apply {
        put("feedId", feedId)
        put("sessionId", sessionId)
        put("slotKey", slotKey)
    }

    companion object {
        const val NATIVE_TESTER_ID = "ca-app-pub-3940256099942544/2247696110"
        private const val MAX_ADS_PER_FEED = 3
    }
}
